utils-intervalrdd/src/main/scala/org/bdgenomics/utils/interval/array/IntervalArray.scala

```diff
@@ -65,7 +67,32 @@ object IntervalArray extends Serializable {
       }).sortByKey()
         .collect
 
-    new IntervalArray(sortedArray, accum.value, sorted = true)
+    buildFn(sortedArray, accum.value)
+  }
+
+  /**
+   * Sorts the RDD and collects it to build an IntervalArray, a sorted array that searches
+   * over ranges. This is used for a left side of the broadcast region join in ADAM.
+   *
+   * @param rdd RDD to build a IntervalArray from.
+   * @return The IntervalArray built from this RDD.
+   */
+  def apply[K <: Interval[K]: ClassTag, T: ClassTag](
+    rdd: RDD[(K, T)]): IntervalArray[K, T] = {
+
+    apply(rdd, apply(_, _, sorted = true))
+  }
+
+  def apply[K <: Interval[K]: ClassTag, T: ClassTag](
+    array: Array[(K, T)],
+    maxIntervalWidth: Long,
+    sorted: Boolean = true): IntervalArray[K, T] = {
+    val sortedArray = if (sorted) {
+      array
+    } else {
+      array.sortBy(_._1)
+    }
+    new ConcreteIntervalArray(sortedArray, maxIntervalWidth)
```

Comment:
I don't like `FooImpl` either.  I usually try hard to find some identifying characteristic; the data structure used, for example, so that you end up with things like `ArrayListFoo`, similar to how the JDK and Guava name things.  Here we have a scala Array with some functionality wrapped around it, but `ArrayIntervalArray` sounds pretty dumb, so I guess I'm ok with `ConcreteIntervalArray`.